tools/base/fpgainfo-c/sysinfo.c

```diff
@@ -0,0 +1,526 @@
+// POSSIBILITY OF SUCH DAMAGE.
+
+#include <errno.h>
+#include <malloc.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <dirent.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <inttypes.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+
+#include "safe_string/safe_string.h"
+#include "opae/fpga.h"
+#include "fpgainfo.h"
+#include "sysinfo.h"
+
+static fpga_result read_string(char *buf, unsigned int buf_size, const char *path)
+{
+	int b = 0;
+	int res = 0;
+	int fd;
+
+	fd = open(path, O_RDONLY);
+	if (fd < 0) {
+		FPGA_MSG("open(%s) failed", path);
+		return FPGA_NOT_FOUND;
+	}
+
+	if ((off_t)-1 == lseek(fd, 0, SEEK_SET)) {
+		FPGA_MSG("seek failed");
+		goto out_close;
+	}
+
+	do {
+		res = read(fd, buf + b, buf_size - b);
+		if (res <= 0) {
+			FPGA_MSG("Read from %s failed", path);
+			goto out_close;
+		}
+		b += res;
+		if (((unsigned)b > buf_size) || (b <= 0)) {
+			FPGA_MSG("Unexpected size reading from %s", path);
+			goto out_close;
+		}
+	} while (buf[b - 1] != '\n' && buf[b - 1] != '\0'
+		 && (unsigned)b < buf_size);
+
+	// erase \n
+	buf[b - 1] = 0;
+
+	close(fd);
+	return FPGA_OK;
+
+out_close:
+	close(fd);
+	return FPGA_OK;
+}
+
+fpga_result fpgainfo_sysfs_read_guid(const char *path, fpga_guid guid)
+{
+	char buf[SYSFS_PATH_MAX];
+	fpga_result fres = FPGA_OK;
+
+	int i;
+	char tmp;
+	unsigned octet;
+
+	if (path == NULL) {
+		FPGA_ERR("Invalid input path");
+		return FPGA_INVALID_PARAM;
+	}
+
+	fres = read_string(buf, sizeof(buf), path);
+	if (FPGA_OK != fres) {
+		return fres;
+	}
+
+	for (i = 0; i < 32; i += 2) {
+		tmp = buf[i + 2];
+		buf[i + 2] = 0;
+
+		octet = 0;
+		sscanf_s_u(&buf[i], "%x", &octet);
+		guid[i / 2] = (uint8_t)octet;
+
+		buf[i + 2] = tmp;
+	}
+
+	return FPGA_OK;
+}
+
+fpga_result fpgainfo_sysfs_read_int(const char *path, int *i)
+{
+	fpga_result fres = FPGA_OK;
+	char buf[SYSFS_PATH_MAX];
+
+	if (path == NULL) {
+		FPGA_ERR("Invalid input path");
+		return FPGA_INVALID_PARAM;
+	}
+
+	fres = read_string(buf, sizeof(buf), path);
+	if (FPGA_OK != fres) {
+		return fres;
+	}
+
+	*i = atoi(buf);
+
+	return FPGA_OK;
+}
+
+fpga_result fpgainfo_sysfs_read_u32(const char *path, uint32_t *u)
+{
+	fpga_result fres = FPGA_OK;
+	char buf[SYSFS_PATH_MAX];
+
+	if (path == NULL) {
+		FPGA_ERR("Invalid input path");
+		return FPGA_INVALID_PARAM;
+	}
+
+	fres = read_string(buf, sizeof(buf), path);
+	if (FPGA_OK != fres) {
+		return fres;
+	}
+
+	*u = strtoul(buf, NULL, 0);
+
+	return FPGA_OK;
+}
+
+fpga_result fpgainfo_sysfs_read_u64(const char *path, uint64_t *u)
+{
+	fpga_result fres = FPGA_OK;
+	char buf[SYSFS_PATH_MAX] = {0};
+
+	if (path == NULL) {
+		FPGA_ERR("Invalid input path");
+		return FPGA_INVALID_PARAM;
+	}
+
+	fres = read_string(buf, sizeof(buf), path);
+	if (FPGA_OK != fres) {
+		return fres;
+	}
+
+	*u = strtoull(buf, NULL, 0);
+
+	return FPGA_OK;
+}
+
+static struct dev_list *fpgainfo_add_dev(const char *sysfspath,
```

Comment:
This is replicating the enumeration code from OPAE.  Is this version DCP-specific?

Reply:
Nope.  I needed access to the sysfs path.  If you have another method for getting it (with fpga_properties?) I'd love to hear it.